File: app/services/nlp_service.py
import spacy
import dateparser
import json
from datetime import datetime
from openai import OpenAI
from app.config import OPENAI_API_KEY
from datetime import datetime

# nlp = spacy.load("en_core_web_sm")
try:
    nlp = spacy.load("en_core_web_sm")
except:
    import os
    os.system("python -m spacy download en_core_web_sm")
    nlp = spacy.load("en_core_web_sm")
client = OpenAI(api_key=OPENAI_API_KEY)


# 🔥 MAIN ENTRY
def analyze_email_thread(messages: list[str]):
    if not messages:
        return empty_response()

    combined_text = "\n---\n".join(messages)

    ai_result = analyze_with_ai(combined_text)

    if ai_result:
        return normalize_response(ai_result)

    return analyze_with_spacy(combined_text)


# 🔹 AI PROCESSING (FIXED)
def analyze_with_ai(text: str):
    try:
        today = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

        response = client.chat.completions.create(
            model="gpt-4.1-mini",
            messages=[
                {
                    "role": "system",
                    "content": f"""
You are an advanced email intelligence system.

Today's datetime: {today}

Context:
- Messages are ordered from oldest → latest
- The LAST message contains the FINAL decision

Rules:
- Always prioritize the latest message
- Interpret relative dates like "tomorrow", "next Monday" using today's date
- Extract only actionable tasks
- Extract time if mentioned (e.g., "5 pm")
- Ignore promotional emails

Return STRICT JSON:
{{
  "task": string or null,
  "deadline": ISO datetime or null,
  "priority": 1-5,
  "summary": string,
  "confidence": 0-1
}}

If no task:
{{
  "task": null,
  "deadline": null,
  "priority": 1,
  "summary": "No actionable task",
  "confidence": 0.5
}}
"""
                },
                {
                    "role": "user",
                    "content": text
                }
            ],
            temperature=0.2
        )

        content = response.choices[0].message.content
        return json.loads(content)

    except Exception as e:
        print("⚠️ AI failed:", e)
        return None


# 🔹 SPACY FALLBACK (FIXED)
def analyze_with_spacy(text: str):
    doc = nlp(text)

    deadline = None

    for ent in doc.ents:
        if ent.label_ == "DATE":
            parsed_date = dateparser.parse(
                ent.text,
                settings={
                    "PREFER_DATES_FROM": "future",
                    "RELATIVE_BASE": datetime.now(),
                    "RETURN_AS_TIMEZONE_AWARE": False
                }
            )
            if parsed_date:
                deadline = parsed_date

    # 🔥 HANDLE DEFAULT TIME (if only date)
    if deadline and deadline.hour == 0:
        deadline = deadline.replace(hour=17)  # assume 5 PM

    return normalize_response({
        "task": "Derived from conversation",
        "deadline": deadline.isoformat() if deadline else None,
        "priority": 2,
        "summary": text[:100],
        "confidence": 0.5
    })


def compute_priority(task: str, deadline: str):
    score = 1  # base

    now = datetime.now()

    # 🔹 DEADLINE BASED
    if deadline:
        try:
            d = datetime.fromisoformat(deadline)
            diff_hours = (d - now).total_seconds() / 3600

            if diff_hours < 0:
                return 5  # expired = critical

            elif diff_hours <= 6:
                score += 4  # 🔥 very urgent

            elif diff_hours <= 24:
                score += 3

            elif diff_hours <= 72:
                score += 2

            else:
                score += 1

        except:
            pass

    # 🔹 KEYWORD BOOST
    text = (task or "").lower()

    high_keywords = [
        "submit", "deadline", "urgent", "asap",
        "important", "immediately", "due"
    ]

    medium_keywords = [
        "meeting", "schedule", "review",
        "prepare", "join", "attend"
    ]

    if any(k in text for k in high_keywords):
        score += 2

    elif any(k in text for k in medium_keywords):
        score += 1

    # 🔹 NORMALIZE (1–5)
    return max(1, min(5, score))

# 🔥 NORMALIZATION (SAFE)
def normalize_response(data):
    task = data.get("task")
    deadline = normalize_deadline(data.get("deadline"))

    # 🔥 override AI priority with our logic
    priority = compute_priority(task, deadline)

    return {
        "task": task,
        "deadline": deadline,
        "priority": priority,
        "summary": data.get("summary", ""),
        "confidence": float(data.get("confidence", 0.5))
    }


# 🔥 DEADLINE NORMALIZER (VERY IMPORTANT)
def normalize_deadline(deadline):
    if not deadline:
        return None

    try:
        # if already ISO
        if isinstance(deadline, str):
            parsed = dateparser.parse(
                deadline,
                settings={
                    "PREFER_DATES_FROM": "future",
                    "RELATIVE_BASE": datetime.now()
                }
            )
        else:
            parsed = deadline

        if not parsed:
            return None

        # default time fix
        if parsed.hour == 0:
            parsed = parsed.replace(hour=17)

        return parsed.isoformat()

    except:
        return None


# 🔥 EMPTY RESPONSE
def empty_response():
    return {
        "task": None,
        "deadline": None,
        "priority": 1,
        "summary": "No actionable task",
        "confidence": 0.0
    }

File: backend/src/services/nlp.service.js
import OpenAI from 'openai';
import axios from 'axios';
import { OPENAI_API_KEY, SPACY_SERVICE_URL } from '../config.js';
import { withRetry } from '../utils/http.utils.js';

const client = new OpenAI({ apiKey: OPENAI_API_KEY });

function restoreRedactions(result, redactions = {}) {
  const restore = value => {
    if (typeof value !== 'string') return value;
    return Object.entries(redactions).reduce(
      (restored, [placeholder, original]) => restored.replaceAll(placeholder, original),
      value,
    );
  };

  return {
    ...result,
    task: restore(result.task),
    summary: restore(result.summary),
    source_snippet: restore(result.source_snippet),
    reasoning: restore(result.reasoning),
  };
}

function buildSanitizedSearchText(result, sanitizedText) {
  // Only text that has already been sanitized is sent to the embeddings API.
  // `result` here preserves placeholders, before identifiers are restored for
  // the user-facing dashboard.
  return [
    `Task: ${result.task || ''}`,
    `Summary: ${result.summary || ''}`,
    `Email: ${sanitizedText.slice(0, 600)}`,
  ].join('\n');
}

async function sanitizeForExternalAI(text) {
  try {
    const res = await withRetry(() => axios.post(`${SPACY_SERVICE_URL}/sanitize-email`, { text }));
    if (typeof res.data?.sanitized_text !== 'string') {
      throw new Error('Privacy service returned no sanitized text');
    }
    return {
      text: res.data.sanitized_text,
      redactions: res.data.redactions || {},
    };
  } catch (err) {
    // Fail closed: never send raw email text to an external AI provider when
    // the privacy service is unavailable or returns an invalid payload.
    console.warn('Email sanitization failed; using local fallback only:', err.message);
    return null;
  }
}

export function computePriority(task, deadline) {
  let score = 1;

  if (deadline) {
    try {
      const diffHours = (new Date(deadline) - new Date()) / 3_600_000;
      if (diffHours < 0) return 5;
      else if (diffHours <= 6) score += 4;
      else if (diffHours <= 24) score += 3;
      else if (diffHours <= 72) score += 2;
      else score += 1;
    } catch { /* ignore invalid dates */ }
  }

  const text = (task || '').toLowerCase();
  const highKw = ['submit', 'deadline', 'urgent', 'asap', 'important', 'immediately', 'due'];
  const medKw = ['meeting', 'schedule', 'review', 'prepare', 'join', 'attend'];

  if (highKw.some(k => text.includes(k))) score += 2;
  else if (medKw.some(k => text.includes(k))) score += 1;

  return Math.max(1, Math.min(5, score));
}

function normalizeResponse(data) {
  const task = data.task ?? null;
  const deadline = data.deadline ?? null;
  return {
    task,
    deadline,
    priority: computePriority(task, deadline),
    summary: data.summary || '',
    confidence: parseFloat(data.confidence ?? 0.5),
    source_snippet: data.source_snippet || null,
    reasoning: data.reasoning || null,
    deadline_source: data.deadline_source ?? (deadline ? 'llm' : null),
  };
}

async function analyzeWithAI(text) {
  const today = new Date().toISOString().replace('T', ' ').slice(0, 19);
  try {
    const res = await client.chat.completions.create({
      model: 'gpt-4.1-mini',
      messages: [
        {
          role: 'system',
          content: `You are an advanced email intelligence system.

Today's datetime: ${today}

Context:
- Messages are ordered from oldest → latest
- The LAST message contains the FINAL decision

Rules:
- Always prioritize the latest message
- Interpret relative dates like "tomorrow", "next Monday" using today's date
- Extract only actionable tasks
- Extract time if mentioned (e.g., "5 pm")
- Ignore promotional emails
- source_snippet must be a short verbatim quote from the email that justifies the task
  (not a paraphrase) -- this is shown to the user as evidence for the extraction
- reasoning must be one sentence explaining why this was flagged as a task
- The email may contain placeholders such as [PERSON_1] or [EMAIL_1]. Preserve those
  placeholders exactly; never attempt to infer the original value.

Return STRICT JSON:
{"task": string or null, "deadline": ISO datetime or null, "priority": 1-5, "summary": string, "confidence": 0-1, "source_snippet": string or null, "reasoning": string or null}

If no task: {"task": null, "deadline": null, "priority": 1, "summary": "No actionable task", "confidence": 0.5, "source_snippet": null, "reasoning": null}`,
        },
        { role: 'user', content: text },
      ],
      temperature: 0.2,
    });

    return JSON.parse(res.choices[0].message.content);
  } catch (err) {
    console.warn('AI analysis failed:', err.message);
    return null;
  }
}

async function analyzeWithSpacy(text) {
  try {
    const res = await withRetry(() => axios.post(`${SPACY_SERVICE_URL}/spacy-analyze`, { text }));
    return res.data;
  } catch (err) {
    console.warn('spaCy service unavailable:', err.message);
    return {
      task: 'Derived from conversation',
      deadline: null,
      priority: 2,
      summary: text.slice(0, 100),
      confidence: 0.5,
      source_snippet: null,
      reasoning: null,
      deadline_source: null,
    };
  }
}

export async function analyzeEmailThread(messages) {
  if (!messages?.length) {
    return { task: null, deadline: null, priority: 1, summary: 'No actionable task', confidence: 0.0 };
  }

  const combinedText = messages.join('\n---\n');
  const privacySafeInput = await sanitizeForExternalAI(combinedText);
  if (!privacySafeInput) {
    return normalizeResponse(await analyzeWithSpacy(combinedText));
  }

  const aiResult = await analyzeWithAI(privacySafeInput.text);
  if (aiResult) {
    const sanitizedResult = normalizeResponse(aiResult);
    return {
      ...restoreRedactions(sanitizedResult, privacySafeInput.redactions),
      // Internal-only: routes remove this before returning a task to the browser.
      _sanitized_search_text: buildSanitizedSearchText(sanitizedResult, privacySafeInput.text),
    };
  }

  const fallbackResult = normalizeResponse(await analyzeWithSpacy(combinedText));
  return {
    ...fallbackResult,
    // Fallback summaries can contain raw email text, so do not embed them.
    _sanitized_search_text: `Email: ${privacySafeInput.text.slice(0, 600)}`,
  };
}
